The great refactor pt1

synsypa opened the pull request:
# Summary

This PR is mainly to reorganize the package so that we can incrementally refactor the quotebot in future PRs.

# Overview
This aspect of the refactor contains:

- Creates a more sane directory structure for everything
    - `src` contains all the code: `src/bot` is bot related items, `src/author_model` is author model related
    - `bin` contains all the data: most notably the AuthorNet checkpoint
- `discord_quote.py` should now **only* *include bot commands. This should both make the main bot script cleaner (it should primarily import relevant functions and tools as well as make it clearer what functions interact which what different quote calls
    - message processing or otherwise uncategorized utiltiy functions exist in `src/bot/utils.py`
    - database interaction utilities exist in `src/bot/db.py`
    - webhook quote functions exist in `src/bot/webhook_quote.py`
    - traditional quote functions exist in `src/bot/quote.py`
- `utils.py` contains generic helper functions for parsing requests and for cleaning up requests, since we do this often across commands
- Additionally:
    - Moves all legacy functions to `legacy_discord_quote.py`
    - Creates a minimal `discord_quote.py` (mostly functional)
    - Replace `logging` with `logzero`

## To-Do
- [x] ~~No command in `discord_quote.py` should manually implement the logic to choose between webhook vs traditional quoting. there should be a separate abstraction (e.g. `def quote()`)of all the logic in i.e. https://github.com/play4honor/discord_quote_bot/blob/8c70f6530ad71bef96ae38cae85fabe64e180dc7/src/bot/discord_quote.py#L181-L199~~
    - The design of this will be tackled in a separate PR and has been moved to issue #85 
- [x] ~~Similarly, it may be better design to abstract as much logic into separate functions as possible for ease of iteration. e.g. rather than having `!misquote` construct the entire message in that bot function, have it call `construct_direct_misquote` and `construct_predicted_misquote` and just return a single string that can be passed to `quote`~~
    - ~~The paradigm i'm thinking here is basically a functional programming one~~
    - The design of this will be tackled in a separate PR and has been moved to issue #85 
- [x] The Dockerfile needs to be changed to properly reflect the new file structure (maybe)
- [x] Many of the utility functions send things to the logger, but none of them actually receive the logger as an argument, i'm not sure what the correct design pattern here is. My instinct is to let pylint be upset, since I don't think it makes sense to have to pass the logger back and forth
    - (See https://github.com/play4honor/discord_quote_bot/pull/84#issuecomment-697260340)
- [x] NONE OF THIS IS TESTED
    - All functionality except for `!misquote` is minimally working
- [x] ~~### REALLY THOUGH ITS NOT TESTED AT ALL~~
- [x] Setup your own personal testing bot (see the README)
    - You need to be able to test your changes. I do not want to be your tester.
- [x] Restore misquote functionality
    - I've commented out both the import and the legacy function for now, because it was not importing in my venv.
- [x] Fix the `parse_request()` function
    - I jerry-rigged this so that it wouldn't give me errors but there is some list/string confusion (which is causing the spacing issues.
    - ![image](https://user-images.githubusercontent.com/4514597/93996047-9c4c3700-fd46-11ea-9899-c4535b12be40.png)
- [x] Am I importing the packages in a correct way? https://github.com/play4honor/discord_quote_bot/pull/84/files#r493388979

## Pre-Merge Checklist

- [x] Does it work in a personal testing bot?
- [x] Has the docker-file paths been edited?

synsypa committed:
Merge pull request #60 from play4honor/development

Deploying URL-quote feature

zhangchuck committed:
Merge pull request #64 from play4honor/development

Deploy cross-channel quoting to production

synsypa committed:
Merge pull request #69 from play4honor/development

Deploy feature/image_quoting

synsypa committed:
Merge pull request #74 from play4honor/development

Deploy features/yoshi_pins

zhangchuck committed:
Merge pull request #81 from play4honor/development

Yoshi Pins Delete Feature and Get/List Improvements

synsypa committed:
(WIP) new directory structure)

synsypa committed:
(WIP) new directory structure)

synsypa committed:
discord_quote.py should only contain bot commands, everything else lives separately

synsypa committed:
discord_quote.py should only contain bot commands, everything else lives separately

synsypa committed:
fixed AuthorNet import

synsypa committed:
fixed AuthorNet import

synsypa committed:
robuster model checkpoint loading

synsypa committed:
robuster model checkpoint loading

zhangchuck commented:
- I'd call the `quote()` function something a bit different. Maybe "speak()" or some variant of "send()". Whenever the bot is *sending* a message, it should call this function rather than any of the myriad ways we can send via the bot.
     **Suggested function signature**:
     **f(`guild_id`, `channel_id`, `message`, `embeds` (optional)) --> Boolean** (success indicator)
- What should be the function signature of the abstract `parse_command()` function?
     **Suggested function signature**: 
     **f(`ctx`, `invoked_command`, `string`) --> Dict(Guild, Channel, Author, OriginMsg,TargetMsg, Reply, InvokedCommand)**
    `invoked_command` helps the `parse_command()` function decide how to parse the command

zhangchuck commented:
On second thought, maybe the first step is to *just* implement the de-tangling of files. e.g., just to fix folder-structure

synsypa commented:
the current message parsing is implemented as a barely edited copy/paste from what used to exist:
https://github.com/play4honor/discord_quote_bot/pull/84/files#diff-0d8dd59df080bc85a83a78768d0f4ebeR54-R78

synsypa committed:
added init to src/author_model for imports

synsypa committed:
added init to src/author_model for imports

zhangchuck commented:
We should tag the current version as release v0.1, or something.

zhangchuck commented:
Also, maybe we should move the internals of all of our commands into a `legacy.py` file (and call on the `legacy.quote()` etc.) so we can rewrite each command onto the new architecture separately instead of trying to rewrite all at once..

zhangchuck committed:
Add pyc and db to gitignore

zhangchuck committed:
Add pyc and db to gitignore

zhangchuck committed:
Move all old code to legacy

zhangchuck committed:
Move all old code to legacy

zhangchuck committed:
Remove bot commands from legacy_discord_quote.py and put them into the new discord_quote.py

zhangchuck committed:
Remove bot commands from legacy_discord_quote.py and put them into the new discord_quote.py

zhangchuck committed:
Add logzero to the requirements.txt

zhangchuck committed:
Add logzero to the requirements.txt

zhangchuck committed:
Add a dropin via logzero in files that use logging.

zhangchuck committed:
Add a dropin via logzero in files that use logging.

zhangchuck committed:
Remove the main (this will not be run directly anymore); comment out misquote as it's causing an error; replace logger.

zhangchuck committed:
Remove the main (this will not be run directly anymore); comment out misquote as it's causing an error; replace logger.

zhangchuck committed:
Import legacy_discord_quote.py and restore functionality to test, me and quote

zhangchuck committed:
Import legacy_discord_quote.py and restore functionality to test, me and quote

zhangchuck committed:
Remove dependency on logging and switch to logzero

zhangchuck committed:
Remove dependency on logging and switch to logzero

zhangchuck committed:
Replace logging in db.py

zhangchuck committed:
Replace logging in db.py

zhangchuck committed:
Make all db functions require a bucket argument

zhangchuck committed:
Make all db functions require a bucket argument

zhangchuck committed:
Missing an argument for db_load()

zhangchuck committed:
Missing an argument for db_load()

zhangchuck committed:
Jerry-rig parse_msg to work, for now...

zhangchuck committed:
Jerry-rig parse_msg to work, for now...

zhangchuck committed:
Restore functionality to all of the pin commands

zhangchuck committed:
Restore functionality to all of the pin commands

zhangchuck committed:
Rename put to pin and delete to drop...

zhangchuck committed:
Rename put to pin and delete to drop...

zhangchuck commented:
## Additional Edits (CZ)

- Switched from default `logging` to `logzero` in most files
- `discord_quote.py` has been stripped bare so that it only contains direct commands
    - to maintain functionality (until we replace in the refactor), we have created `legacy_discord_quote.py`, which contains all of the v0.1 versions of the commands
    - `discord_quote.py` imports these functions and (currently) directly passes along the needed arguments
    - functionality is similar but not perfect

zhangchuck commented on src/bot/legacy_discord_quote.py:
```diff
@@ -4,153 +4,27 @@
-streamInstance.setFormatter(fmt)
-log.addHandler(streamInstance)
-log.setLevel(logging.DEBUG)
+#import src.author_model.author_model as author
```
This was failing!

zhangchuck commented on src/bot/legacy_discord_quote.py:
```diff
@@ -294,242 +149,7 @@ async def quote(ctx, *, request:str):
-
-@bot.command()
-async def misquote(ctx , *target : discord.User):
+""" async def misquote(ctx , *target : discord.User):
```
Since the import was failing, I commented out this entire function.

zhangchuck commented on src/bot/discord_quote.py:
```diff
@@ -0,0 +1,138 @@
+
+@bot.command()
+async def me(ctx, *text : str):
+    """
```
Needs a docstring.

zhangchuck commented on src/bot/discord_quote.py:
```diff
@@ -0,0 +1,138 @@
+    await old.quote(ctx, request=request, bot=bot)
+
+@bot.command()
+async def misquote(ctx , *target : discord.User):
```
Needs a docstring, also we need to link this over to `old.misquote()`

zhangchuck commented:
# To-Do's (@synsypa and @DoubleSec)

- [ ] Setup your own personal testing bot (see the README)
    - You need to be able to test your changes. I do not want to be your tester.
- [ ] Restore misquote functionality
    - I've commented out both the import and the legacy function for now, because it was not importing in my venv.
- [ ] Fix the `parse_request()` function
    - I jerry-rigged this so that it wouldn't give me errors but there is some list/string confusion (which is causing the spacing issues.
    - ![image](https://user-images.githubusercontent.com/4514597/93996047-9c4c3700-fd46-11ea-9899-c4535b12be40.png)
- [ ] Edit the docker file need to be edited for this to successfully deploy to `quote-bot-dev`
- [ ] Am I importing the packages in a correct way?

zhangchuck commented on src/bot/discord_quote.py:
```diff
@@ -0,0 +1,138 @@
+import sys
+import os
+
+sys.path.append('../../')
```
I had to do this to jerry-rig it properly. Not sure if it's a venv problem or something else.

Recall that we deploy by doing `python3 discord_quote.py` (our working directory is `src/bot/`.

synsypa replied:
yea, i haven't done the change but i think my plan is to change the dockerfile to use the base directory as its wd, and then deploy with `python3 src/bot/discord_quote.py`. I can make the adjustments.

zhangchuck commented:
```Many of the utility functions send things to the logger, but none of them actually receive the logger as an argument, i'm not sure what the correct design pattern here is. My instinct is to let pylint be upset, since I don't think it makes sense to have to pass the logger back and forth```

There's some weird waterfalling that goes on with logger and not sure exactly how it works. For now, I've added:

```
from logzero import logger as log
```

which allows everything to run fine.

synsypa committed:
fixed model loading, path issue +  fking pickle bs

synsypa committed:
fixed model loading, path issue +  fking pickle bs

synsypa committed:
correctly separated make targets

synsypa committed:
correctly separated make targets

synsypa commented:
i think i fixed the loading thing, it was an issue with a bad file path + some annoying `pickle` behavior from `torch.load()`.

Also fixed up the dockerfile and made some make targets. Didn't have a chance to test yet.

zhangchuck commented on Dockerfile:
```diff
@@ -8,10 +8,14 @@ RUN apt-get -y update && \
 # Copy the latest version of the bot 
-COPY ./discord_quote /discord_quote
+COPY ./src /src
+COPY ./bin /bin
```
You're probably just mirroring what we did before, but any idea why we need to do this step?

zhangchuck committed:
Edited setup.py for new folder structure

zhangchuck committed:
Edited setup.py for new folder structure

synsypa committed:
Merge branch 'the-great-refactor-pt1' of github.com:play4honor/discord_quote_bot into the-great-refactor-pt1

synsypa committed:
reenabled misquote, fixed reply parsing

synsypa committed:
ignore scratchpad notebook

synsypa committed:
makefile targets

synsypa committed:
dockerfile installs module rather than sets pythonpath, makefile has local target

synsypa committed:
removed some hacky bs

synsypa committed:
ignore egg

synsypa commented:
Tested locally off a docker container, everything should be working.

zhangchuck commented on src/bot/discord_quote.py:
```diff
@@ -0,0 +1,142 @@
+import botocore
+
+# Other Packages
+from logzero import logger as log
```
Are we generally okay with this? Or, should we just use

```
from logzero import logger
```

zhangchuck commented on src/bot/webhook_quote.py:
```diff
@@ -2,6 +2,8 @@
 import arrow
 import re
 
+from typing import Union
```
What's this do?

zhangchuck commented:
London Gents Tussle Messily
